pubsub-service-api.yml

```diff
@@ -457,33 +468,35 @@ components:
       format: int32
       minimum: 0
       example: 7
+    MessageProperty:
+      type: object
+      required:
+        - from
+        - data
+      properties:
+        data:
+          description: "The content encoded with base64 standard alphabet with padding."
+          type: string
+          example: "YWJj"
+        from:
+          description: "The peer ID of the sender."
+          type: string
+          example: "QmcXhKGEAJT9DwwzKeEZsWrYTTGkzBdgeKif2MdAXBswJv"
+        pubkey:
+          description: "The sender's public key encoded with base64 standard alphabet with padding in X.509 DER format."
+          type: string
+          format: byte
+          example: "MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAmreiLkzVz2qlc2a4z8FbdUTzvka3vQVqPX0QRqr+Lgb6NOw0RVR8WTySavlg8zjOqgyRcEXMvmL1R34jMped68SBaRUyBBmvhT8ptzOZhZdSB4PPRwAfhILWyHnUm9JBPEmjN+Xi9wNt0WH8ds5S+RRnBoD1BagVxAykzjCJnZzAtKQYiU4jIG8zm7jwaWTZjukKaVsHAY7kjPF/8i2u3u+xltKGXYCwKbg621fTaEKBzC2mQORod9fw0qLhrOT/N1CuC5xCrlqvaDv8J4yKvJMd6XBG3lcsP/yhcCw9FjHdpolsRsomeSofvl7/OdaPMDimpdT012ohZAmPntUyNQIDAQAB"
+        signature:
+          description: "The signature for the message."
```

Comment:
unless I'm missing something this should be required if `from` is required when **publishing**, but wouldn't be when **reading**. Therefore, you need two different property.

Reply:
> but isn't from always required

Yes, and that's my point.

Libp2p's pubsub requires for every message to be signed, which means `from`, `pubkey` (possibly derived from `from`) and `signature` need to match. From there, two options:
- the client pass everything, and the full thing is relayed by the service
- the client pass only the message, and the service's keypair is used (but I'm not quite sure that makes sense in practice)